File: claude/web/src/lib/arcade/rally.ts
/**
 * Rally — a deterministic SteamBench arcade *adversarial vision* env (Pong).
 *
 * TypeScript mirror of `harness/steambench_harness/envs/rally.py` (and the
 * headless twin of the pixel runtime's RallyGame). Drive the left paddle, return
 * the attacking opponent's shots; the run ends the first time you miss.
 *
 * Determinism note: integer dynamics; the only rng is each serve's vertical
 * velocity (one draw). The opponent is a deterministic tracker/attacker, so the
 * whole rally is reproducible from (seed, actions).
 */

import { ActionSpace, Env } from "./base";
import type { AchievementSpec, Action, Observation } from "./types";

const W = 168;
const H = 120;
const PADW = 4;
const PADH = 26;
const AGENT_X = 4;
const OPP_X = W - 4 - PADW;
const PSPEED = 6;
const OPP_SPEED = 7;
const BS = 6;
const BASE_SPEED = 4;
const MAX_SPEED = 8;
const SERVE_VY = [-2, -1, 1, 2];
const ACTIONS = ["up", "down", "stay"] as const;

export class Rally extends Env {
  static env_id = "arcade/rally";
  static appid = 9000016;
  static displayName = "Rally";
  static description =
    "A Pong-style duel: drive the left paddle, read the ball, and beat the " +
    "built-in opponent — score when it can't reach your return, and the run " +
    "ends the first time you miss. Adversarial, temporal and visual at once.";
  static verify_mode = "replay" as const;
  static action_space = new ActionSpace([...ACTIONS]);
  static achievements: AchievementSpec[] = [
    { id: "rally_5", name: "Warm-up", description: "Return 5 of the opponent's shots.", rarity_hint: 0.6 },
    { id: "rally_15", name: "Rally", description: "Return 15 shots.", rarity_hint: 0.3 },
    { id: "rally_30", name: "Backboard", description: "Return 30 shots.", rarity_hint: 0.12 },
    { id: "rally_60", name: "Iron Wall", description: "Return 60 shots.", rarity_hint: 0.035 },
    { id: "rally_120", name: "Untouchable", description: "Return 120 shots.", rarity_hint: 0.006 },
  ];
  private static LADDER: [number, string][] = [
    [5, "rally_5"], [15, "rally_15"], [30, "rally_30"], [60, "rally_60"], [120, "rally_120"],
  ];

  ay: number; oy: number;
  bx: number; by: number; vx: number; vy: number; speed: number;
  points: number; alive: boolean;

  constructor() {
    super();
    this.ay = Math.trunc((H - PADH) / 2);
    this.oy = Math.trunc((H - PADH) / 2);
    this.bx = Math.trunc(W / 2); this.by = Math.trunc(H / 2);
    this.vx = -BASE_SPEED; this.vy = 1; this.speed = BASE_SPEED;
    this.points = 0; this.alive = true;
  }

  private serve(): void {
    this.bx = Math.trunc(W / 2);
    this.by = Math.trunc(H / 2);
    this.speed = BASE_SPEED;
    this.vx = -BASE_SPEED;
    this.vy = SERVE_VY[this.rng.randrange(SERVE_VY.length)];
  }

  reset(seed = 0): Observation {
    this.begin(seed);
    this.ay = Math.trunc((H - PADH) / 2);
    this.oy = Math.trunc((H - PADH) / 2);
    this.points = 0;
    this.alive = true;
    this.serve();
    return this.observe(0.0);
  }

  step(action: Action): Observation {
    if (this.done) return this.observe(0.0);
    const name = this.actionSpace.name(action);
    this.steps += 1;

    if (name === "up") this.ay = Math.max(0, this.ay - PSPEED);
    else if (name === "down") this.ay = Math.min(H - PADH, this.ay + PSPEED);

    const ballCy = this.by + Math.trunc(BS / 2);
    const oppCy = this.oy + Math.trunc(PADH / 2);
    if (ballCy < oppCy - 1) this.oy = Math.max(0, this.oy - OPP_SPEED);
    else if (ballCy > oppCy + 1) this.oy = Math.min(H - PADH, this.oy + OPP_SPEED);

    this.bx += this.vx;
    this.by += this.vy;
    if (this.by <= 0) { this.by = 0; this.vy = -this.vy; }
    else if (this.by >= H - BS) { this.by = H - BS; this.vy = -this.vy; }

    let reward = 0.0;
    if (this.vx < 0 && this.bx <= AGENT_X + PADW && this.bx > AGENT_X - this.speed) {
      if (this.by + BS > this.ay && this.by < this.ay + PADH) {
        this.bx = AGENT_X + PADW;
        this.speed = Math.min(MAX_SPEED, this.speed + 1);
        this.vx = this.speed;
        const off = (this.by + Math.trunc(BS / 2)) - (this.ay + Math.trunc(PADH / 2));
        // Python uses floor division (//) — `off` can be negative, where
        // Math.trunc would diverge — so floor here to stay bit-identical.
        this.vy = Math.max(-7, Math.min(7, Math.floor(off / 2)));
        this.points += 1;
        reward = 1.0;
        for (const [thresh, aid] of Rally.LADDER) {
          if (this.points >= thresh) this.unlock(aid);
        }
      }
    } else if (this.vx > 0 && this.bx + BS >= OPP_X && this.bx + BS < OPP_X + PADW + this.speed) {
      this.bx = OPP_X - BS;
      this.speed = Math.min(MAX_SPEED, this.speed + 1);
      this.vx = -this.speed;
      const agentCy = this.ay + Math.trunc(PADH / 2);
      this.vy = agentCy < Math.trunc(H / 2) ? this.speed - 2 : -(this.speed - 2);
    }

    if (this.bx + BS < 0) {
      this.alive = false;
      this.done = true;
      reward = -1.0;
    } else if (this.bx > W) {
      this.serve();
    }

    this.score = this.points;
    return this.observe(reward);
  }

  legalActions(): string[] {
    return [...ACTIONS];
  }

  render(): string {
    return `points=${this.points} ay=${this.ay} oy=${this.oy} ball=(${this.bx},${this.by}) v=(${this.vx},${this.vy}) ${this.alive ? "ALIVE" : "DEAD"}`;
  }

  private observe(reward: number): Observation {
    return {
      step: this.steps,
      state: {
        agent_x: AGENT_X, agent_y: this.ay, opp_x: OPP_X, opp_y: this.oy,
        paddle_w: PADW, paddle_h: PADH, paddle_speed: PSPEED,
        ball_x: this.bx, ball_y: this.by, ball_vx: this.vx, ball_vy: this.vy, ball_size: BS,
        width: W, height: H, points: this.points, score: this.points, alive: this.alive,
      },
      text: this.render(),
      frame: null,
      legal_actions: this.legalActions(),
      score: this.points,
      done: this.done,
      reward: reward,
      info: { points: this.points, alive: this.alive },
    };
  }
}

File: claude/harness/steambench_harness/envs/rally.py
"""Rally — a deterministic SteamBench arcade *adversarial vision* env (Pong).

The one structural gap the other vision games leave: an opponent. You drive the
left paddle, a built-in AI drives the right, and a ball volleys between you —
speeding up every hit. Win a point when the opponent can't reach your return;
the run ends the first time YOU miss. It folds three things into one: read the
ball's motion (temporal), position to intercept (vision), and out-place a
reacting adversary (adversarial).

Integer dynamics; the only randomness is each serve's vertical velocity (one
draw per serve), so ``(seed, actions)`` traces replay exactly. The opponent is a
deterministic tracker, so it is part of the reproducible dynamics.
"""

from __future__ import annotations

from typing import Union

from ..protocol import AchievementSpec, ActionSpace, Env, Observation, VerifyMode
from ..registry import register

# Shared with runtime/pixel_game.py RallyGame — keep in lock-step (parity test).
W, H = 168, 120
PADW, PADH = 4, 26
AGENT_X = 4                 # left paddle x (the player)
OPP_X = W - 4 - PADW        # right paddle x (the opponent)
PSPEED = 6                  # agent paddle speed
OPP_SPEED = 7               # opponent always reaches — it ATTACKS, aiming away
BS = 6                      # ball size
BASE_SPEED = 4
MAX_SPEED = 8
_SERVE_VY = [-2, -1, 1, 2]
_ACTIONS = ["up", "down", "stay"]


@register
class Rally(Env):
    env_id = "arcade/rally"
    appid = 9000016
    name = "Rally"
    description = (
        "A Pong-style duel: drive the left paddle, read the ball, and beat the "
        "built-in opponent — score when it can't reach your return, and the run "
        "ends the first time you miss. Adversarial, temporal and visual at once."
    )
    verify_mode = VerifyMode.REPLAY
    action_space = ActionSpace(_ACTIONS)
    achievements = [
        AchievementSpec("rally_5", "Warm-up", "Return 5 of the opponent's shots.", 0.6),
        AchievementSpec("rally_15", "Rally", "Return 15 shots.", 0.3),
        AchievementSpec("rally_30", "Backboard", "Return 30 shots.", 0.12),
        AchievementSpec("rally_60", "Iron Wall", "Return 60 shots.", 0.035),
        AchievementSpec("rally_120", "Untouchable", "Return 120 shots.", 0.006),
    ]
    _LADDER = [(5, "rally_5"), (15, "rally_15"), (30, "rally_30"),
               (60, "rally_60"), (120, "rally_120")]

    def __init__(self) -> None:
        super().__init__()
        self.ay = (H - PADH) // 2
        self.oy = (H - PADH) // 2
        self.bx = W // 2
        self.by = H // 2
        self.vx = -BASE_SPEED
        self.vy = 1
        self.speed = BASE_SPEED
        self.points = 0
        self.alive = True

    def _serve(self) -> None:
        self.bx = W // 2
        self.by = H // 2
        self.speed = BASE_SPEED
        self.vx = -BASE_SPEED                       # serve toward the agent
        self.vy = _SERVE_VY[self.rng.randrange(len(_SERVE_VY))]

    def reset(self, seed: int = 0) -> Observation:
        self._begin(seed)
        self.ay = (H - PADH) // 2
        self.oy = (H - PADH) // 2
        self.points = 0
        self.alive = True
        self._serve()
        return self._observe(0.0)

    def step(self, action: Union[int, str]) -> Observation:
        if self.done:
            return self._observe(0.0)
        name = self.action_space.name(action)
        self.steps += 1

        if name == "up":
            self.ay = max(0, self.ay - PSPEED)
        elif name == "down":
            self.ay = min(H - PADH, self.ay + PSPEED)

        # opponent tracks the ball's center vertically (a touch slower)
        ball_cy = self.by + BS // 2
        opp_cy = self.oy + PADH // 2
        if ball_cy < opp_cy - 1:
            self.oy = max(0, self.oy - OPP_SPEED)
        elif ball_cy > opp_cy + 1:
            self.oy = min(H - PADH, self.oy + OPP_SPEED)

        self.bx += self.vx
        self.by += self.vy
        if self.by <= 0:
            self.by = 0
            self.vy = -self.vy
        elif self.by >= H - BS:
            self.by = H - BS
            self.vy = -self.vy

        reward = 0.0
        # agent paddle hit = a successful RETURN (the score); english from the
        # spot it struck (a centred hit goes straight, an edge hit cuts).
        if self.vx < 0 and self.bx <= AGENT_X + PADW and self.bx > AGENT_X - self.speed:
            if self.by + BS > self.ay and self.by < self.ay + PADH:
                self.bx = AGENT_X + PADW
                self.speed = min(MAX_SPEED, self.speed + 1)
                self.vx = self.speed
                off = (self.by + BS // 2) - (self.ay + PADH // 2)
                self.vy = max(-7, min(7, off // 2))
                self.points += 1
                reward = 1.0
                for thresh, aid in self._LADDER:
                    if self.points >= thresh:
                        self._unlock(aid)
        # opponent ATTACKS: it always reaches (OPP_SPEED ≥ any vy), then aims the
        # ball at the half the agent ISN'T in, harder as the rally speeds up.
        elif self.vx > 0 and self.bx + BS >= OPP_X and self.bx + BS < OPP_X + PADW + self.speed:
            self.bx = OPP_X - BS
            self.speed = min(MAX_SPEED, self.speed + 1)
            self.vx = -self.speed
            agent_cy = self.ay + PADH // 2
            self.vy = (self.speed - 2) if agent_cy < H // 2 else -(self.speed - 2)

        if self.bx + BS < 0:                        # agent missed -> run ends
            self.alive = False
            self.done = True
            reward = -1.0
        elif self.bx > W:                           # opponent let one by -> re-serve
            self._serve()

        self.score = self.points
        return self._observe(reward)

    def legal_actions(self) -> list[str]:
        return list(_ACTIONS)

    def render(self) -> str:
        return (f"points={self.points} ay={self.ay} oy={self.oy} ball=({self.bx},{self.by}) "
                f"v=({self.vx},{self.vy}) {'ALIVE' if self.alive else 'DEAD'}")

    def _observe(self, reward: float) -> Observation:
        return Observation(
            step=self.steps,
            state={
                "agent_x": AGENT_X,
                "agent_y": self.ay,
                "opp_x": OPP_X,
                "opp_y": self.oy,
                "paddle_w": PADW,
                "paddle_h": PADH,
                "paddle_speed": PSPEED,
                "ball_x": self.bx,
                "ball_y": self.by,
                "ball_vx": self.vx,
                "ball_vy": self.vy,
                "ball_size": BS,
                "width": W,
                "height": H,
                "points": self.points,
                "score": self.points,
                "alive": self.alive,
            },
            text=self.render(),
            legal_actions=self.legal_actions(),
            score=float(self.points),
            done=self.done,
            reward=float(reward),
            info={"points": self.points, "alive": self.alive},
        )
